### PR TITLE
Set default for AccountContact/AccountInvoice connector_id to 0

### PR DESCRIPTION
**Pull request by mattwire**
- Set default for AccountContact/AccountInvoice connector_id to 0.
- Add a new system check to notify if there are NULL values in connector_id (you have to fix manually because it's quite likely you'll have duplicates in the existing database because the sql constraint doesn't work if you have a mix of 0 and NULL).
- Code cleanup and upgrade civix.
- Add constants for account_status_id.
- Fix day zero setting (was a string is now an array).

### CODE REVIEW

**Comment by mattwire** [comment]
@eileenmcnaughton This should be ready for review/merge if you get time. The NULL/0 thing has proved to be the cause of quite a few sync issues using both Xero and Quickbooks integrations.

**Comment by eileenmcnaughton** [comment]
@mattwire I agree with the principle here - NULL vs 0 is causing problems.
 
Shouldn't we make it required & update all existing to 0? Then it would be done & dusted. I'm assuming the other DAO changes are just what happened when you regenerated as the DAO was not well maintained.

Also - I'm not a fan of constants cos I find them way harder to grep for than using words - but those constants (right at the start) also don't see to relate to anything in the PR (ditto I always use the extension name rather than `E::Short` for grepping reasons - I think the main advantage of the `E::short` is you can create boilerplate code with it but outside boilerplate I priorities 'greppiness')

**Comment by mattwire** [comment]
The constants are in use in my civixero changes eg. here: https://github.com/eileenmcnaughton/nz.co.fuzion.civixero/pull/101/files#diff-ae8087fabc2278332d47c7a8c20ef96d456517ae2f3ecde0f4a3c8e8ac10139aR371

I prefer words too but what we currently have stored in the database are numbers and it's pretty much impossible to grep / understand the code at the moment without cross-referencing the numbers to something. The constants make it much clearer without changing the underlying data structure.

**Comment by mattwire** [comment]
> Shouldn't we make it required & update all existing to 0? Then it would be done & dusted.

Yes. But two different systems had lots of entries which would become duplicates and therefore note cleanly upgrade because it was not being consistently set. So I went for the option of forcing a manual update.

> I'm assuming the other DAO changes are just what happened when you regenerated as the DAO was not well maintained.

Yes there should be no other deliberate changes.

**Comment by eileenmcnaughton** [comment]
@mattwire ah - we should define a pseudoconstant function for that field then from the looks of it - then we can specify the full array - ie name, icon, description, etc (or leave some for later)

**Comment by eileenmcnaughton** [comment]
@mattwire I needed to regen the DAO to add in some html attributes so I included the default of 0 in there as well (so hopefully you can drop the schema changes part from this & it will be done)

Re the upgrade - what about doing the UPGRADE with a join - ie upgrade them if they won't clash - that way next round we could say 'they've had long enough, let's make it required'

File: accountsync.php [hunk 1]
```diff
@@ -348,7 +348,14 @@ function _accountsync_get_skip_invoice_create_entities($connector_id) {
  * @throws \CiviCRM_API3_Exception
  */
 function _accountsync_get_invoice_day_zero(int $connector_id): ?string {
-  return _accountsync_get_entity_action_settings($connector_id)['account_sync_contribution_day_zero'] ?? NULL;
+  $dayZero = _accountsync_get_entity_action_settings($connector_id)['account_sync_contribution_day_zero'] ?? NULL;
+  if (is_array($dayZero)) {
+    // Before https://github.com/eileenmcnaughton/nz.co.fuzion.accountsync/commit/dc5aa75523433e71ce09118fae2e59dc1eeb398b
```
Comment (eileenmcnaughton) [inline accountsync.php:353]: Ug that's awful - I feel like it's a core bug - ie it seems to be core deciding (incorrectly) it is an array?

**Comment by eileenmcnaughton** [comment]
After seeing what the 'clear error' action was actually doing I've added new schema fields to fix those. Since I was doing an upgrader I pulled in your upgrader lines too - I also added an attempt to update as many of the NULL connector_ids as seems safe

File: accountsync.php (hunk 1 above)
Comment (mattwire) [reply accountsync.php:353]: It was the change from `html_type` = 'date' to `html_type` = 'Date' I think. Definitely got saved as string before and array afterwards when you save settings.

Comment (eileenmcnaughton) [reply accountsync.php:353]: Aha - 

image

**Comment by mattwire** [comment]
Thanks, closing as mostly merged

**Pull request closed without merging**